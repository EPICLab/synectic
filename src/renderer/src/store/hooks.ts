import { createAsyncThunk } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

export const createAppAsyncThunk = createAsyncThunk.withTypes<{
  state: RootState;
  dispatch: AppDispatch;
  rejectValue: string;
}>();

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
