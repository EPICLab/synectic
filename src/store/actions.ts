import { UUID, Repository, Card, Stack, Metafile, Filetype } from '../types';

export enum ActionKeys {
  INITIALIZE_CANVAS,
  ADD_CARD, REMOVE_CARD, UPDATE_CARD,
  ADD_STACK, REMOVE_STACK, UPDATE_STACK,
  ADD_FILETYPE, REMOVE_FILETYPE, UPDATE_FILETYPE,
  ADD_METAFILE, REMOVE_METAFILE, UPDATE_METAFILE,
  ADD_REPO, REMOVE_REPO, UPDATE_REPO
}

export type Actions =
  | InitializeCanvasAction
  | AddCardAction
  | RemoveCardAction
  | UpdateCardAction
  | AddStackAction
  | RemoveStackAction
  | UpdateStackAction
  | AddFiletypeAction
  | RemoveFiletypeAction
  | UpdateFiletypeAction
  | AddMetafileAction
  | RemoveMetafileAction
  | UpdateMetafileAction
  | AddRepoAction
  | RemoveRepoAction
  | UpdateRepoAction;

type InitializeCanvasAction = {
  type: ActionKeys.INITIALIZE_CANVAS;
};

type AddCardAction = {
  type: ActionKeys.ADD_CARD;
  id: UUID;
  card: Card;
};

type RemoveCardAction = {
  type: ActionKeys.REMOVE_CARD;
  id: UUID;
};

type UpdateCardAction = {
  type: ActionKeys.UPDATE_CARD;
  id: UUID;
  card: Partial<Card>;
};

type AddStackAction = {
  type: ActionKeys.ADD_STACK;
  id: UUID;
  stack: Stack;
}

type RemoveStackAction = {
  type: ActionKeys.REMOVE_STACK;
  id: UUID;
};

type UpdateStackAction = {
  type: ActionKeys.UPDATE_STACK;
  id: UUID;
  stack: Partial<Stack>;
};

type AddFiletypeAction = {
  type: ActionKeys.ADD_FILETYPE;
  id: UUID;
  filetype: Filetype;
};

type RemoveFiletypeAction = {
  type: ActionKeys.REMOVE_FILETYPE;
  id: UUID;
};

type UpdateFiletypeAction = {
  type: ActionKeys.UPDATE_FILETYPE;
  id: UUID;
  filetype: Partial<Filetype>;
};

type AddMetafileAction = {
  type: ActionKeys.ADD_METAFILE;
  id: UUID;
  metafile: Metafile;
};

type RemoveMetafileAction = {
  type: ActionKeys.REMOVE_METAFILE;
  id: UUID;
};

type UpdateMetafileAction = {
  type: ActionKeys.UPDATE_METAFILE;
  id: UUID;
  metafile: Partial<Metafile>;
};

type AddRepoAction = {
  type: ActionKeys.ADD_REPO;
  id: UUID;
  repo: Repository;
};

type RemoveRepoAction = {
  type: ActionKeys.REMOVE_REPO;
  id: UUID;
}

type UpdateRepoAction = {
  type: ActionKeys.UPDATE_REPO;
  id: UUID;
  repo: Partial<Repository>;
};