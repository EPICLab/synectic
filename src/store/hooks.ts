import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

export type AppThunkAPI<R = unknown> = { state: RootState, dispatch: AppDispatch, rejectValue: R };

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;