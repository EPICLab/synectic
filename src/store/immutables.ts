/**
 * Immutably copy the values of all enumerable own properties from old object 
 * and map new values onto it in a new target object.
 * @param oldObject The initial source object from which to copy properties.
 * @param newValues The new value object from which to copy properties.
 * @return New object containing all enumerable own properties from both params.
 */
export const updateObject = <T>(oldObject: T, newValues: Partial<T>): T => {
  return Object.assign({}, oldObject, newValues);
};

/**
 * Immutably append a new item to an array by constructing a new combined array.
 * @param oldArray The initial source Array object.
 * @param newItem A new item to be appended to array.
 * @returns New array containing all items from old array and including new item.
 */
export const addItemInArray = <T>(oldArray: T[], newItem: T): T[] => [...oldArray, newItem];

/**
 * Immutably remove an element from an array by producing a new reduced array.
 * @param array The initial source Array object.
 * @param item A target item contained within the array.
 * @returns New array containing all elements from array excluding target item.
 */
export const removeItemInArray = (array: string[], item: string): string[] => {
  return array.filter(element => element !== item);
}

/**
 * Immutably append a new item to a map by constructing a new combined map.
 * @param map The initial source key-value map object.
 * @param newItem A new item to be appended to map.
 * @returns New map containing all items from old map and including new item.
 */
export const addItemInMap = <T extends { id: string }>(map: { [id: string]: T }, newItem: T): { [id: string]: T } => {
  const updatedItems: { [id: string]: T } = {};
  for (const k in map) {
    updatedItems[k] = map[k];
  }
  updatedItems[newItem.id] = newItem;
  return updatedItems;
};

/**
 * Immutably remove an item from a map by producing a new reduced map.
 * @param map The initial source key-value map object.
 * @param itemId An id associated with an item contained in the map.
 * @returns New map containing all items from map excluding item with matching id.
 */
export const removeItemInMap = <T extends { id: string }>(map: { [id: string]: T }, itemId: string): { [id: string]: T } => {
  return Object.keys(map).reduce((items: { [id: string]: T }, id) => {
    if (id !== itemId) {
      items[id] = map[id];
    }
    return items;
  }, {});
}

/**
 * Immutably remove items from map using specified filter function.
 * @param map The initial source key-value map object.
 * @param filterFn Filter function that returns true for each item in map that 
 * meets conditions specified in function.
 */
export const removeMatchesInMap = <T>(map: { [id: string]: T }, filterFn: (item: T) => boolean): { [id: string]: T } => {
  return Object.keys(map).reduce((items: { [id: string]: T }, id) => {
    if (!filterFn(map[id])) {
      items[id] = map[id];
    }
    return items;
  }, {});
};

/**
 * Immutably filters items from map using specified filter function and applies a 
 * callback function to update each matching item.
 * @param map The initial source key-value map object.
 * @param filterFn Filter function that returns true for each item in map that 
 * meets conditions specified in function.
 * @param updateItemCallback Callback function to apply towards items that meet 
 * filter function predicates.
 */
export const updateMatchesInMap = <T>(map: { [id: string]: T }, filterFn: (item: T) => boolean, updateItemCallback: (item: T) => T): { [id: string]: T } => {
  const updatedItems: { [id: string]: T } = {};
  for (const k in map) {
    if (filterFn(map[k])) {
      updatedItems[k] = updateItemCallback(map[k]);
    } else {
      updatedItems[k] = map[k];
    }
  }
  return updatedItems;
};

/**
 * Immutably update a specific item in a key-value map based on item id 
 * and applying a callback function to that item.
 * @param map The initial source key-value map object.
 * @param itemId An id associated with an item contained in the map.
 * @param updateItemCallback Callback function to apply towards item with matching id.
 * @returns New map containing all items from map including updated item.
 */
export const updateItemInMapById = <T extends { id: string }>(map: { [id: string]: T }, itemId: string, updateItemCallback: (item: T) => T): { [id: string]: T } => {
  return updateMatchesInMap(map, (item => item.id === itemId), updateItemCallback);
};

