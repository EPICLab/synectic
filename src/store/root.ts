import { combineReducers } from 'redux';
import { canvasReducer } from './reducers/canvas';
import { stackReducer } from './reducers/stacks';
import { cardReducer } from './reducers/cards';
import { filetypeReducer } from './reducers/filetypes';
import { metafileReducer } from './reducers/metafiles';
import { repoReducer } from './reducers/repos';
import { modalReducer } from './reducers/modal';

export const rootReducer = combineReducers({
  canvas: canvasReducer,
  stacks: stackReducer,
  cards: cardReducer,
  filetypes: filetypeReducer,
  metafiles: metafileReducer,
  repos: repoReducer,
  modals: modalReducer
});

export type RootState = ReturnType<typeof rootReducer>;