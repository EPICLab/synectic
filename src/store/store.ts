import { combineReducers, configureStore } from '@reduxjs/toolkit';
import stacksReducer from './slices/stacks';
import cardsReducer from './slices/cards';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import reposReducer from './slices/repos';
import modalsReducer from './slices/modals';
import { FLUSH, PAUSE, PERSIST, persistReducer, persistStore, PURGE, REGISTER, REHYDRATE } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

export const rootReducer = combineReducers({
    stacks: stacksReducer,
    cards: cardsReducer,
    filetypes: filetypesReducer,
    metafiles: metafilesReducer,
    repos: reposReducer,
    modals: modalsReducer
});

const persistConfig = {
    key: 'root',
    storage: storage,
    version: 0,
    blacklist: ['filetypes']
}

export const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) => [
        ...getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
            }
        }),
    ]
});
const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default ({ store: store, persistor: persistor });