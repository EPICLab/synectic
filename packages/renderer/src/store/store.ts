import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
