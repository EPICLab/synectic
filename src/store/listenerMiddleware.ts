import { createListenerMiddleware, addListener, TypedStartListening, TypedAddListener, isRejected } from '@reduxjs/toolkit';
import { DateTime } from 'luxon';
import cacheSelectors from './selectors/cache';
import metafileSelectors from './selectors/metafiles';
import { cacheRemoved } from './slices/cache';
import { cardAdded, cardRemoved, cardUpdated } from './slices/cards';
import { isDirectoryMetafile, isFilebasedMetafile, isFileMetafile, metafileUpdated } from './slices/metafiles';
import { RootState, AppDispatch } from './store';
import { checkoutBranch } from './thunks/branches';
import { subscribe, unsubscribe } from './thunks/cache';
import { createMetafile, fetchMetafile, fetchParentMetafile, updatedVersionedMetafile, updateFilebasedMetafile } from './thunks/metafiles';
import { fetchRepo, createRepo } from './thunks/repos';

export const listenerMiddleware = createListenerMiddleware<RootState>();

export type AppStartListening = TypedStartListening<RootState, AppDispatch>;

export const startAppListening = listenerMiddleware.startListening as AppStartListening;

export const addAppListener = addListener as TypedAddListener<RootState, AppDispatch>;

const isRejectedAction = isRejected(fetchMetafile, createMetafile, updateFilebasedMetafile, updatedVersionedMetafile, fetchParentMetafile, fetchRepo, createRepo);

startAppListening({
    matcher: isRejectedAction,
    effect: async (action) => {
        console.group(`${action.type} : ${DateTime.local().toHTTP()}`);
        console.log(action.error);
        console.groupEnd();
    }
});

startAppListening({
    matcher: updatedVersionedMetafile.pending.match,
    effect: async (action, listenerApi) => {
        listenerApi.dispatch(metafileUpdated({ ...action.meta.arg, loading: [...action.meta.arg.loading, 'versioned'] }));
    }
});

startAppListening({
    matcher: updatedVersionedMetafile.fulfilled.match,
    effect: async (action, listenerApi) => {
        listenerApi.dispatch(metafileUpdated({ ...action.meta.arg, loading: action.meta.arg.loading.filter(flag => flag !== 'versioned') }));
    }
});

startAppListening({
    matcher: checkoutBranch.pending.match,
    effect: async (action, listenerApi) => {
        const metafile = metafileSelectors.selectById(listenerApi.getState(), action.meta.arg.metafileId);
        if (metafile) listenerApi.dispatch(metafileUpdated({ ...metafile, loading: [...metafile.loading, 'checkout'] }));
    }
});

startAppListening({
    matcher: checkoutBranch.fulfilled.match,
    effect: async (action, listenerApi) => {
        const metafile = metafileSelectors.selectById(listenerApi.getState(), action.meta.arg.metafileId);
        if (metafile) listenerApi.dispatch(metafileUpdated({ ...metafile, loading: metafile.loading.filter(flag => flag !== 'checkout') }));
    }
});

startAppListening({
    predicate: (action, _, previousState) => {
        if (cardAdded.match(action)) return true;
        if (cardUpdated.match(action)) {
            const prev = previousState.cards.entities[action.payload.id];
            return prev ? action.payload.metafile !== prev.metafile : false;
        }
        return false;
    },
    effect: async (action, listenerApi) => {
        const metafile = metafileSelectors.selectById(listenerApi.getState(), action.payload.metafile);

        if (metafile && (cardAdded.match(action) || cardUpdated.match(action))) {

            if (isFilebasedMetafile(metafile)) {
                const updated = await listenerApi.dispatch(updateFilebasedMetafile(metafile)).unwrap();
                await listenerApi.dispatch(updatedVersionedMetafile(metafile));

                if (isDirectoryMetafile(updated)) {
                    const children = metafileSelectors.selectByIds(listenerApi.getState(), updated.contains);
                    await Promise.all(children.filter(isFilebasedMetafile).map(child =>
                        listenerApi.dispatch(updateFilebasedMetafile(child))
                    ));
                }

                if (isFileMetafile(updated)) {
                    await listenerApi.dispatch(subscribe({ path: updated.path.toString(), metafile: updated.id }));
                }
            }
        }
    },
});

startAppListening({
    actionCreator: cardRemoved,
    effect: async (action, listenerApi) => {
        const state = listenerApi.getState();
        const card = listenerApi.getOriginalState().cards.entities[action.payload];
        const metafile = card ? state.metafiles.entities[card.metafile] : undefined;
        if (card && metafile) {
            const childIds = isDirectoryMetafile(metafile) ? metafile.contains :
                isFileMetafile(metafile) ? [metafile.id] : [];
            const metafiles = metafileSelectors.selectByIds(state, childIds);
            metafiles.filter(isFilebasedMetafile).forEach(metafile => {
                const existing = cacheSelectors.selectById(state, metafile.path.toString());
                if (existing) {
                    if (existing && existing.reserved.length > 1) {
                        listenerApi.dispatch(unsubscribe({ path: existing.path, metafile: metafile.id }));
                    } else {
                        listenerApi.dispatch(cacheRemoved(existing.path));
                    }
                }
            });
        }
    }
});