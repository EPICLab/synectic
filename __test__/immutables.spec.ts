import * as utils from '../src/store/immutables';

type State = {
  config: { id: string; display: boolean; enabled: boolean };
  colors: string[];
  cats: { [id: string]: { id: string; age: number; color: string } };
}

const state: State = {
  config: {
    id: 'test_config',
    display: false,
    enabled: false
  },
  colors: ['red', 'green', 'blue'],
  cats: {
    'pearl': { id: 'pearl', age: 3, color: 'brown' },
    'mittens': { id: 'mittens', age: 11, color: 'white' },
    'scruffy': { id: 'scruffy', age: 6, color: 'yellow' }
  }
}

describe('immutables.updateObject', () => {
  it('updateObject immutably replaces target object', () => {
    const newConfig = { id: 'updated_config', display: true, enabled: true };
    const newState = utils.updateObject(state, { config: newConfig });
    expect(state.config).not.toMatchObject(newConfig);
    expect(newState.config).toMatchObject(newConfig);
  });

  it('updateObject immutably updates only specified properties', () => {
    const newState = utils.updateObject(state, { config: { ...state.config, enabled: true } });
    expect(newState.config.id).toMatch(state.config.id);
    expect(newState.config.enabled).toBe(true);
  });
});

describe('immutables.addItemInArray', () => {
  it('addItemInArray immutably adds object to array', () => {
    const newColors = utils.addItemInArray(state.colors, 'yellow');
    expect(newColors).toHaveLength(4);
    expect(newColors).toContain('yellow');
  });
});

describe('immutables.removeItemInArray', () => {
  it('removeItemInArray immutably removes object from array', () => {
    const newColors = utils.removeItemInArray(state.colors, 'green');
    expect(newColors).toHaveLength(2);
    expect(newColors).toMatchSnapshot();
  });

  it('removeItemInArray resolves non-matching item and returns original state', () => {
    const newColors = utils.removeItemInArray(state.colors, 'purple');
    expect(newColors).toMatchObject(state.colors);
  });
});

describe('immutables.addItemInMap', () => {
  it('addItemInMap immutably adds object to map', () => {
    const oscar = { id: 'oscar', age: 5, color: 'black' };
    const newCats = utils.addItemInMap(state.cats, oscar);
    expect(Object.keys(newCats)).toHaveLength(4);
    expect(newCats['oscar']).toBeTruthy();
  });
});

describe('immutables.removeItemInMap', () => {
  it('removeItemInMap immutably removes object from map', () => {
    const newCats = utils.removeItemInMap(state.cats, 'mittens');
    expect(Object.keys(newCats)).toHaveLength(2);
    expect(newCats['mittens']).toBeFalsy();
  });

  it('removeItemInMap resolves non-matching item and returnS original state', () => {
    const newCats = utils.removeItemInMap(state.cats, 'tom');
    expect(newCats).toMatchObject(state.cats);
  });
});

describe('immutables.updateMatchesInMap', () => {
  it('updateMatchesInMap immutably filters and updates only specified properties of matching items', () => {
    const elderCats = utils.updateMatchesInMap(state.cats, (cat => cat.age > 10), cat => {
      return utils.updateObject(cat, { ...cat, id: 'old ' + cat.id });
    });
    expect(elderCats).toMatchSnapshot();
  });
});

describe('immutables.updateItemInMapById', () => {
  it('updateItemInMapById immutably updates only specified properties of matching item', () => {
    const agedCats = utils.updateItemInMapById(state.cats, 'scruffy', cat => {
      return utils.updateObject(cat, { ...cat, age: cat.age + 1 });
    });
    expect(agedCats['scruffy']).not.toMatchObject(state.cats['scruffy']);
    expect(agedCats).toMatchSnapshot();
  });
});