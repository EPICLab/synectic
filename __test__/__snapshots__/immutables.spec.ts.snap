// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`immutables.removeItemInArray removeItemInArray immutably removes object from array 1`] = `
Array [
  "red",
  "blue",
]
`;

exports[`immutables.updateItemInMapById updateItemInMapById immutably updates only specified properties of matching item 1`] = `
Object {
  "mittens": Object {
    "age": 11,
    "color": "white",
    "id": "mittens",
  },
  "pearl": Object {
    "age": 3,
    "color": "brown",
    "id": "pearl",
  },
  "scruffy": Object {
    "age": 7,
    "color": "yellow",
    "id": "scruffy",
  },
}
`;

exports[`immutables.updateMatchesInMap updateMatchesInMap immutably filters and updates only specified properties of matching items 1`] = `
Object {
  "mittens": Object {
    "age": 11,
    "color": "white",
    "id": "old mittens",
  },
  "pearl": Object {
    "age": 3,
    "color": "brown",
    "id": "pearl",
  },
  "scruffy": Object {
    "age": 6,
    "color": "yellow",
    "id": "scruffy",
  },
}
`;
